File: backend/legacy/valor-ingreso.service.js
const axios = require("axios");

const config = require("../config/env");

const {
  createHttpsAgent
} = require("../utils/tls");

const {
  DatacreditoAuthService
} = require("./datacredito-auth.service");

const logger = require("../utils/logger");

class ValorIngresoService {
  constructor() {
    this.auth =
      new DatacreditoAuthService(
        "income"
      );

    this.httpsAgent =
      createHttpsAgent();
  }

  async consultar(params) {
    const token =
      await this.auth.getAccessToken();

    const headers = {
      access_token: token,

      client_id:
        config.datacredito
          .income
          .clientId,

      client_secret:
        config.datacredito
          .income
          .clientSecret,

      Accept:
        "application/json",

      "Accept-Encoding":
        "gzip,deflate"
    };

    try {
      const response =
        await axios.get(
          config.datacredito
            .income
            .serviceUrl,

          {
            params,

            timeout:
              config.datacredito
                .timeout,

            httpsAgent:
              this.httpsAgent,

            headers
          }
        );

      logger.info(
        "Consulta Valor Ingreso ejecutada",
        {
          status:
            response.status
        }
      );

      return {
        success: true,
        status: response.status,
        data: response.data
      };
    } catch (err) {
      if (
        err.response?.status === 401 ||
        err.response?.status === 403
      ) {
        this.auth.clearToken();

        const newToken =
          await this.auth.getAccessToken();

        const retryResponse =
          await axios.get(
            config.datacredito
              .income
              .serviceUrl,

            {
              params,

              timeout:
                config.datacredito
                  .timeout,

              httpsAgent:
                this.httpsAgent,

              headers: {
                ...headers,
                access_token:
                  newToken
              }
            }
          );

        return {
          success: true,
          status:
            retryResponse.status,
          data:
            retryResponse.data
        };
      }

      logger.error(
        "Error consumiendo Valor Ingreso",
        {
          status:
            err.response?.status,

          response:
            err.response?.data,

          message:
            err.message
        }
      );

      throw err;
    }
  }
}

module.exports =
  ValorIngresoService;

File: backend/services/valor_ingreso_service.py
import os

import httpx

from backend.services.datacredito_auth import (
    DatacreditoAuthService,
)
from backend.utils.loggers import get_logger
from backend.utils.tls import load_certificates


logger = get_logger("datacredito.income")


class ValorIngresoService:

    def __init__(self) -> None:

        self.auth = DatacreditoAuthService(
            "income"
        )

        self.url = os.getenv(
            "DATACREDITO_INCOME_SERVICE_URL"
        )

    async def consultar(self, params: dict):

        if not self.url:
            raise RuntimeError(
                "Falta la variable "
                "DATACREDITO_INCOME_SERVICE_URL"
            )

        token = await self.auth.get_access_token()

        headers = {
            "access_token": token,
            "client_id": os.getenv(
                "DATACREDITO_INCOME_CLIENT_ID"
            ),
            "client_secret": os.getenv(
                "DATACREDITO_INCOME_CLIENT_SECRET"
            ),
            "Accept": "application/json",
            "Accept-Encoding": "gzip,deflate",
        }

        verify = load_certificates()

        timeout = (
            float(
                os.getenv(
                    "DATACREDITO_TIMEOUT",
                    "30000",
                )
            )
            / 1000
        )

        try:

            async with httpx.AsyncClient(
                verify=verify,
                timeout=timeout,
            ) as client:

                response = await client.get(
                    self.url,
                    params=params,
                    headers=headers,
                )

            if response.status_code in (401, 403):

                logger.warning(
                    "Token de Valor Ingreso rechazado. "
                    "Renovando token."
                )

                self.auth.clear_token()

                token = await self.auth.get_access_token()

                headers["access_token"] = token

                async with httpx.AsyncClient(
                    verify=verify,
                    timeout=timeout,
                ) as client:

                    response = await client.get(
                        self.url,
                        params=params,
                        headers=headers,
                    )

            response.raise_for_status()

            return {
                "success": True,
                "status": response.status_code,
                "data": response.json(),
            }

        except httpx.HTTPError as error:

            logger.error(
                "Error consultando Valor Ingreso: %s",
                error,
            )

            raise RuntimeError(
                "Error consumiendo servicio "
                "Valor Ingreso"
            ) from error